### PR TITLE
Parallelize stdout and clean up PersistentWorker

### PR DESCRIPTION
**Pull request by jeffzoch**
For discussion in #494

### CODE REVIEW

File: src/main/kotlin/io/bazel/worker/PersistentWorker.kt [hunk 1]
```diff
@@ -50,81 +60,66 @@ class PersistentWorker(
...
-        return@runBlocking action()
-      }
-    }
-  }
+  val scope = CoroutineScope(Job() + Dispatchers.IO)
 
   @ExperimentalCoroutinesApi
   override fun start(execute: Work) = WorkerContext.run {
+    //Use channel to serialize writing output
+    val writeChannel = Channel<WorkerProtocol.WorkResponse>(UNLIMITED)
```
Comment (jeffzoch) [inline src/main/kotlin/io/bazel/worker/PersistentWorker.kt:68]: Main idea is to launch the producer in its own coroutine and have the collector run to completion synchronously. The producer does the compilation in parallel and emits results across a channel which is serially read from. Flows are serial by default so we don't need to synchronize access on the consumer side.

File: src/main/kotlin/io/bazel/worker/PersistentWorker.kt [hunk 2]
```diff
@@ -50,81 +60,66 @@ class PersistentWorker(
...
-        blockable {
+      runBlocking {
+        launch {
           generateSequence { WorkRequest.parseDelimitedFrom(io.input) }
-        }.asFlow()
-          .map { request ->
-            info { "received req: ${request.requestId}" }
-            async {
-              doTask("request ${request.requestId}") { ctx ->
+            .asFlow()
```
Comment (jdai8) [inline src/main/kotlin/io/bazel/worker/PersistentWorker.kt:73]: Is this flow necessary? Would it be simpler to do something like:
```
generateSequence { WorkRequest.parseDelimitedFrom(io.input) }.forEach {
  launch(Dispatchers.DEFAULT) {
    // do the work
    writeChannel.send(response)
  }
}

Comment (jeffzoch) [reply src/main/kotlin/io/bazel/worker/PersistentWorker.kt:73]: @jdai8 yeah that would totally work too - I didn't want to rock the boat too much but I can change it to that as its easier to reason about.

Comment (jeffzoch) [reply src/main/kotlin/io/bazel/worker/PersistentWorker.kt]: Just went ahead and changed to just use coroutines (very similar to your thread implementation now just using kotlin primitives I guess. I'd think we would need to be compiling thousands of files to *maybe* even see a perf difference)

**Review by restingbull — APPROVED** [review]

**Comment by restingbull** [comment]
Much cleaner, thank you. 

Also educational. 👍

**Pull request merged**